src/plugins/share/config.ts

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema, TypeOf } from '@kbn/config-schema';
+
+export const configSchema = schema.object({
+  hideEmbed: schema.conditional(
```

Comment:
__note:__ Up to you, but I'm wondering if it'd be easier\cleaner to rely on [`env.buildFlavor`](https://github.com/elastic/kibana/pull/161930) provided by the Core (both on the server and client side) instead of creating a new, technically immutable, config flag.

Reply:
It doesn't have to be a config value.  You can expose an API on the `start` contract and call it from the relevant `serverless` plugin.

Flexing based on `is(serverless)` means you lose context on related serverless decisions.  It will be hard to parse out what code applies to a particular functional decision if everything is keyed off of that binary proposition.

Reply:
>  It will be hard to parse out what code applies to a particular functional decision if everything is keyed off of that binary proposition.

Well, it might sound good in theory, but I'm not sure it's the case in practice. We have discussed this a lot internally for the past few weeks, so sorry for lots of text :see_no_evil: 

The things like `share.hideEmbed` just add an unnecessary level of indirection and complexity around something that is only relevant to the Serverless offering. As far as I am aware, there is no valid use case for "hiding embed" controls, no such "feature" is on our roadmap, and might never be. We're just trying to come up with a workaround for the limitations we have in one specific offering, pretending that it's not just an "isServerless" flag. Moreover, we might lift some of these limitations eventually, making many of these "fake-future" flags unnecessary.

When I read the code and see `share.hideEmbed` and friends, unlike `isServerless`, the following questions pop up in my mind:

* What use case is this for?
* What consumers outside of this plugin rely on it?
* Are there any consumers at all?
* Can I safely change or remove it if I have to?
* Since it's exposed through a contract, can I rely on it in my other plugin? If not, why?

No matter how you expose it, via config or via plugin contracts, these are all public APIs that we're supposed to support backward compatibility for. It's just that we usually ignore breaking changes in our programmatic API contracts since we don't have many external plugins that could be affected by these changes.

In summary, what I want to say is that we have at least three different use cases when it comes to Serverless, and trying to solve them with one tool might not be the best approach:

1. Kibana functionality that should not be available or work differently in Serverless but represents a well-isolated "domain" that makes sense on its own, and there is a high chance that this functionality might be disabled/re-configured in other contexts/offerings in the future. This is the perfect use case for feature config flags and additional APIs on the setup/start contracts.

2. Kibana functionality that mostly depends on counterpart Elasticsearch functionality that should not be available or work differently in Serverless. In this case, Kibana shouldn't care about Serverless directly and ideally should rely on available Elasticsearch capabilities exposed in one way or another. If Elasticsearch doesn't _yet_ expose a way to advertise certain capabilities, relying on `isServerless` in Kibana as an interim solution sounds like the best\fairest approach.

3. Kibana functionality which customizations **only** make sense in the Serverless context. Unless we really can tick all the boxes from the option 1, I don't think we should pretend that it's a "feature" and just use the `isServerless` flag until we feel it's indeed a feature so that we can drop `isServerles`s and expose a proper feature flag or programmatic feature management API.

Having said that, it's not my call here, just wanted to share my perspective!

Reply:
> Kibana functionality which customizations only make sense in the Serverless context. Unless we really can tick all the boxes from the option 1, I don't think we should pretend that it's a "feature" and just use the isServerless flag until we feel it's indeed a feature so that we can drop isServerless and expose a proper feature flag or programmatic feature management API.

Is there an example that uses `isServerless` flag that you can point me too? I have no strong opinions about implementation path.

Reply:
The [`env.buildFlavor`](https://github.com/elastic/kibana/pull/161930), provided by the Core (both on the server and client side), is a relatively new addition, but we're planning to use it like that on the server side https://github.com/elastic/kibana/pull/162087/files#diff-4fbb178a08b40c270265f939de2e735ba79ea1181ea5c1ebca18dd032fcfcf2eR98, and I see Maps is using this flag on the client side as well: https://github.com/elastic/kibana/blob/490c34c053e6342a7f08d00879770145c2d1cb95/src/plugins/maps_ems/public/lazy_load_bundle/create_ems_client.ts#L24-L26

Reply:
Thanks for the examples.

As another data point, https://github.com/elastic/kibana/pull/161528, mentioned in AppEx updates 2023-08-02, uses a yaml configuration to disable functionality in serverless. I don't think there is a consistent approach across Kibana

Reply:
I would +1 to the yaml config option as it allows us to flip this switch regardless of serverless. I.e. we might want to hide this functionality in all offerings in the future.